File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/TopInterview2Round/removeNthFromEnd_19.java
package algo.tzashinorpu.SpecialSubject.TopInterview2Round;

import algo.tzashinorpu.SpecialSubject.ListNode;

public class removeNthFromEnd_19 {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        // fast 要删除的下一个节点
        // slow 要删除的上一个节点
//        ListNode fast = head, slow = head;
        ListNode dummyNode = new ListNode();
        dummyNode.next = head;
        ListNode slow = dummyNode;
        ListNode fast = dummyNode;
        while (n-- != -1) fast = fast.next;
        while (fast != null) {
            fast = fast.next;
            slow = slow.next;
        }
        slow.next = slow.next.next;
        return dummyNode.next;
    }
}

File: Clion/labuladong/chapter0/linklist/removeNthFromEnd_19.cpp
//
// Created by TzashiNorpu on 5/31/2023.
//
#include "../../header/header.h"

using namespace labuladong;

ListNode *Solution::removeNthFromEnd(ListNode *head, int n) {
  if (!head)return nullptr;
  auto *dummy = new ListNode(-1);
  dummy->next = head;
  ListNode *r = dummy;
  ListNode *l = dummy;
  while (n--) {
    r = r->next;
  }
  while (r && r->next) {
    r = r->next;
    l = l->next;
  }
  l->next = l->next->next;
  return dummy->next;
}